### DISCUSSION TITLE
How setuptools finds Visual Studio compiler

### DISCUSSION THREAD

**Discussion by Mark531** [pypa/setuptools#4091] [Q&A] [answered]
Hello,

I have a question about the setuptools package: how does it find the Visual Studio compiler when it compiles a C++ extension on Windows? From the documentation (https://setuptools.pypa.io/en/latest/userguide/ext_modules.html), I understand that the compiler is looked for in the sysconfig variables CC, CXX or CSHARED, or in the environment variables CC, CPP or CXX.

Though, when I call `python -m sysconfig` from an Anaconda prompt, I don't see any of the interesting variables and when I call `set` from a VS Developer Command Prompt, I don't see any of the interesting environment variables either. So, how does it work?

Thanks in advance,
Mark

**Comment by abravalheri** [comment, marked as answer]
Setuptools uses `vswhere` to find modern versions of the toolchain.

https://github.com/pypa/setuptools/blob/main/setuptools/msvc.py

**Reply by Mark531** [reply]
Thanks for pointing the source file that handles it. I see the code is very specific to Microsoft VS... So, first it looks for vswhere.exe then executes it to find the file vcvarsall.bat to know the last version and then it executes the batch file, then it dumps the environment variables... and I'm lost. :) How complex...
